pkg/dashboard/resource/apigateway.go

```diff
@@ -319,71 +344,58 @@ func (agr *apiGatewayResource) getAPIGatewayInfoFromRequest(request *http.Reques
 	}
 
 	if err = agr.processAPIGatewayInfo(&apiGatewayInfoInstance,
-		nameRequired,
 		specRequired,
 		request.Header.Get("x-nuclio-project-name")); err != nil {
 
-		return nil, nuclio.WrapErrBadRequest(errors.Wrap(err, "Failed to process api gateway info"))
+		return nil, errors.Wrap(err, "Failed to process api gateway info")
 	}
 
 	return &apiGatewayInfoInstance, nil
 }
 
-func (agr *apiGatewayResource) processAPIGatewayInfo(apiGatewayInfoInstance *apiGatewayInfo,
-	nameRequired bool,
-	specRequired bool,
-	projectName string) error {
-	var err error
-
-	if apiGatewayInfoInstance.Meta == nil {
-		return nuclio.NewErrBadRequest("Api gateway must be provided with metadata")
-	}
-
-	// enrich meta name from spec if needed
-	if apiGatewayInfoInstance.Meta.Name == "" {
-		apiGatewayInfoInstance.Meta.Name = apiGatewayInfoInstance.Spec.Name
-	}
+func (agr *apiGatewayResource) enrichAPIGatewayInfo(apiGatewayInfoInstance *apiGatewayInfo, projectName string) error {
 
-	// override namespace if applicable
-	if apiGatewayInfoInstance.Meta != nil {
-		apiGatewayInfoInstance.Meta.Namespace = agr.getNamespaceOrDefault(apiGatewayInfoInstance.Meta.Namespace)
+	// ensure spec exists if it's required
+	if apiGatewayInfoInstance.Spec == nil {
+		apiGatewayInfoInstance.Spec = &platform.APIGatewaySpec{}
 	}
 
-	// meta must exist
-	if apiGatewayInfoInstance.Meta == nil ||
-		(nameRequired && apiGatewayInfoInstance.Meta.Name == "") ||
-		apiGatewayInfoInstance.Meta.Namespace == "" {
-
-		if nameRequired {
-			err = errors.New("Api gateway name and namespace must be provided in metadata")
-		} else {
-			err = errors.New("Api gateway namespace must be provided in metadata")
+	// enrich project name when specified
+	if projectName != "" {
+		if apiGatewayInfoInstance.Meta.Labels == nil {
+			apiGatewayInfoInstance.Meta.Labels = map[string]string{}
 		}
 
-		return nuclio.WrapErrBadRequest(err)
+		apiGatewayInfoInstance.Meta.Labels["nuclio.io/project-name"] = projectName
 	}
 
-	// ensure spec exists if it's required
-	if apiGatewayInfoInstance.Spec == nil {
-		if specRequired {
-			err = errors.New("Api gateway spec must be provided")
-			return nuclio.WrapErrBadRequest(err)
-		}
-
-		apiGatewayInfoInstance.Spec = &platform.APIGatewaySpec{}
-	}
+	// override namespace if applicable
+	apiGatewayInfoInstance.Meta.Namespace = agr.getNamespaceOrDefault(apiGatewayInfoInstance.Meta.Namespace)
 
 	// status is optional, ensure it exists
 	if apiGatewayInfoInstance.Status == nil {
 		apiGatewayInfoInstance.Status = &platform.APIGatewayStatus{}
 	}
 
-	if projectName != "" {
-		if apiGatewayInfoInstance.Meta.Labels == nil {
-			apiGatewayInfoInstance.Meta.Labels = map[string]string{}
-		}
+	return nil
+}
 
-		apiGatewayInfoInstance.Meta.Labels["nuclio.io/project-name"] = projectName
+func (agr *apiGatewayResource) processAPIGatewayInfo(apiGatewayInfoInstance *apiGatewayInfo,
```

Comment:
In this case, the validation is only basic, making sure spec and meta exists, otherwise we won't be able to even enrich it - because it is nil, and if we will initialize it, we won't be able to to return a better error saying "meta/spec wasn't provided"
The actual inner validations are on Platform (shared across platforms - nuctl/dashboard)

Reply:
specRequired == false on DELETE and true otherwise

Reply:
Why do we want to inform user "spec" is missing? what if we will enrich it with an empty spec? thenthe validation will say "spec.X field is missing". that would save you the trouble of checking spec is nill and specRequired boolean

(true for meta as well)